include/bbp/sonata/compartment_sets.h

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <bbp/sonata/nodes.h>
+
+namespace bbp {
+namespace sonata {
+namespace detail {
+class CompartmentLocation;
+class CompartmentSetFilteredIterator;
+class CompartmentSet;
+class CompartmentSets;
+}  // namespace detail
+/**
+ * CompartmentLocation public API.
+ * 
+ * This class uniquely identifies a compartment by a set of gid, section_idx and offset:
+ * 
+ * - gid: Global ID of the cell (Neuron) to which the compartment belongs. No 
+ * overlaps among populations.
+ * - section_idx: Absolute section index. Progressive index that uniquely identifies the section.
+ *  There is a mapping between neuron section names (i.e. dend[10]) and this index.
+ * - offset: Offset of the compartment along the section. The offset is a value between 0 and 1
+ */
+class SONATA_API CompartmentLocation
+{
+  public:
+    CompartmentLocation();
+    CompartmentLocation(const int64_t gid, const int64_t section_idx, const double offset);
+    explicit CompartmentLocation(const std::string& content);
+    explicit CompartmentLocation(std::unique_ptr<detail::CompartmentLocation>&& impl);
+    CompartmentLocation(const CompartmentLocation& other);
+    CompartmentLocation& operator=(const CompartmentLocation& other);
+    CompartmentLocation(CompartmentLocation&&) noexcept;
+    CompartmentLocation& operator=(CompartmentLocation&&) noexcept;
+    ~CompartmentLocation();
+
+    bool operator==(const CompartmentLocation& other) const noexcept;
+    bool operator!=(const CompartmentLocation& other) const noexcept;
+
+    uint64_t gid() const;
+    uint64_t sectionIdx() const;
+    double offset() const;
+
+    std::string toJSON() const;
+
+  private:
+    std::unique_ptr<detail::CompartmentLocation> impl_;
+};
+
+class SONATA_API CompartmentSetFilteredIterator {
```

Comment:
I'm not %100 convince this iterator was necessary; yes, there is some memory savings, but I don't expect the compartment sets to become so large that keeping things simple wasn't the best choice.

We strive to keep libsonata as small as possible, without adding features unless people request them and as they turn up.
We have limited people to maintain it, so we try to limit as much as possible all the features that we need to maintain.

That being said, since it's here, I guess that ship has already sailed.

Reply:
in this way the user can decide to use the iterator and not place a new vector of selections or place the vector with the filter function if memory is a contraint.

It is true that is not strictly necessary. It can also be ditched easily. There is 1 function that uses this in CompartmentSet and we can use filter and values() to effectively get the same result (with more memory consumption). 

If you want I can add a small comment on how to easily remove this. I thought it was a nice addition for the user. Also, considering that it is a free function (not nested inside CompartmentSet) is easily removable if we want

Reply:
> If you want I can add a small comment on how to easily remove this.

Naw, either it stays or it goes.

> I thought it was a nice addition for the user.

I think you're right; so we'll keep it based on the fact that doing it manually requires `quite` a bit more memory.